Add support for Gamma regression

yangs16 opened the pull request:
Add the support for gamma regression and also add a metric of gamma-nloglik.

Known problem: some of the first iterations will return trees with only a root node or shallow trees (max_depth is smaller than the specified one). Need helps from the community to solve this problem.

See the running log snippet below:
-------------------------------------------------------------------------------------------
tree pruning end, 1 roots, 6 extra nodes, 0 pruned nodes, max_depth=3
 tree pruning end, 1 roots, 0 extra nodes, 0 pruned nodes, max_depth=0
 tree pruning end, 1 roots, 4 extra nodes, 0 pruned nodes, max_depth=2
 tree pruning end, 1 roots, 2 extra nodes, 0 pruned nodes, max_depth=1
 tree pruning end, 1 roots, 2 extra nodes, 0 pruned nodes, max_depth=1
 tree pruning end, 1 roots, 8 extra nodes, 0 pruned nodes, max_depth=3
 tree pruning end, 1 roots, 10 extra nodes, 0 pruned nodes, max_depth=3
 tree pruning end, 1 roots, 10 extra nodes, 0 pruned nodes, max_depth=3
 tree pruning end, 1 roots, 10 extra nodes, 0 pruned nodes, max_depth=3
-------------------------------------------------------------------------------------------

yangs16 commented:
Currently only support the gamma regression with log link.

yangs16 commented:
After adding a constant to the un-transformed predictions, now the resulting tree in each round will have the max_depth as specified.

yangs16 commented:
Can somebody give some comments on this pull request? Currently, the results given the same dataset is just a bit better than R's GLM. Anybody knows how to further improve? According to my experience, the Poission regression with XGBoost is about 10% better than R'S GLM.

khotilov commented:
I gave it a few runs with some simulated gamma-response data with a few distribution shapes - and it seemed to work reasonably well! 

By the way, the existing `base_score` parameter (default value is 0.5) could be used instead of adding a new one - `base_score` is a global transformed bias. DMatrix also has a `base_margin` parameter which allows to set local untransformed (WRT "margin") initial offsets for data points.

And, by chance, may you be possibly willing to look into whether this framework could be extended to Tweedie distributions with power parameter between (1,2) (Gamma is a Tweedie with power=2)? For not too small distribution shapes, Gamma could often be reasonably approximated by OLS against log-transformed response, but 1-to-2 Tweedies are more versatile. I started looking into that a while ago, but didn't ever have time to figure that out.

tqchen commented:
Sorry for the delayed response. As @khotilov  suggested, it is likely we can re-use base_score parameter as un-transformed bias. Previously for logistic regression, there is a transformation function that transform back to the base_score as untransformed score.

yangs16 commented:
Thanks @khotilov @tqchen for the comments. I have removed the newly added bias parameter and the global base_score works fine. As for Tweedie implementation, I have no plan in near future. I wish I can work on that later.

tqchen commented on src/learner.cc:
```diff
@@ -168,6 +168,12 @@ class LearnerImpl : public Learner {
       cfg_["max_delta_step"] = "0.7";
     }
 
+    if (cfg_.count("lp_bias") == 0 &&
```
please remove this block

tqchen commented:
A few more things:
- Please remove the lp_bias configuration block
- Please add a test-cases of gamma regression to python or R API
- It would be great if you can create a markdown(rmarkdown if you are using R) to demonstrate how this feature can be used

yangs16 commented:
Thanks @tqchen for the great comments. I have removed the config block for lp_bias and added a demo for running gamma regression in Python.

tqchen commented:
Last comment, I would like to avoid adding new datasets to the repo as it makes it growing larger. It would be nice if you can provide a script to download the data as in other demos.

Thanks

yangs16 commented:
Thanks @tqchen. I have removed the autoclaims datasets from the demo/data directory and added a R script for generating the dataset. Thanks again for all the helpful comments and instructions.
